### PR TITLE
Fix open issues/PRs #197 #108 #111 #236 #238

### LINKED ISSUE #197: Remove battery API option when Firefox 52 is released
As of Firefox 52, the battery API has been removed: https://bugzilla.mozilla.org/show_bug.cgi?id=1313580

Once 52 is out (it will replace ESR 45 too), there's no need to have this in user.js anymore.

### LINKED ISSUE #235: Change value of security.ask_for_password

### LINKED ISSUE #236: Write descriptions to the rest of the prefs
* [ ] webgl.disable-extensions
* [ ] webgl.disable-fail-if-major-performance-caveat
* [ ] webgl.enable-debug-renderer-info
* [ ] media.gmp-*
* [ ] browser.aboutHomeSnippets.updateUrl
* [ ] network.stricttransportsecurity.preloadlist
* [x] WebIDE

### LINKED ISSUE #238: Move 'Known problems' to NOTICE: fields alongside relevant prefs
... and generate a _Known problems and limitations_ README section from `NOTICE:` fields (improve `gen-readme.sh`)

Example output:

 * `browser.frames.enabled: false` breaks websites relying on IFRAMEs
 * `privacy.clearOnShutdown.passwords: true` will cause **all your saved passwords** to be removed
 * `network.cookie.cookieBehavior: 1` breaks a number of payment gateways so you may need to comment it out
 * ...

Same reasoning and advantages as #229

### PR DESCRIPTION
**Pull request by nodiscc**
~~Incomplete PR in the current state (#238 work still going on) but the~~ first commits [can be reviewed/merged independently](https://gist.github.com/piscisaureus/3342247) (see commit messages):

* add firefox version indication for battery api (fixes #197)
* add commented out prefs for *effective* default search engine choice (closes #108)
* document disabling push notifications (references in commit message, closes #111)
* Add missing preferences descriptions (Fixes #236)
 * set security.ask_for_password to 2 (lock password storage periodically), set password storage lock timeout to 1 minute (fixes #235)
* improve and reorder FAQ 			d1e4b6d
* rename LICENSE.txt file to LICENSE, link to it from README 		eb45fd7
* more explicit NOTICE: fields 			581309b
* gen-readme.sh: add 'known problems and limitations' section generation 			930967b
* run make, generate 'kwown problems and limitations' readme section, fixes #238  b70f602

### CODE REVIEW

**Comment by nodiscc** [comment]
Done. Check out https://github.com/nodiscc/user.js/tree/2017-mar#known-problems-and-limitations

**Comment by nodiscc** [comment]
@pyllyukko ping, waiting for review.

**Comment by pyllyukko** [comment]
> @pyllyukko ping, waiting for review.

Pong. Sorry about that. I've been quite busy this week and have had zero time to look into these. Will get working on these ASAP, but it might still take few days.

**Comment by pyllyukko** [comment]
I seem to be unable to figure out how the review functionality works :stuck_out_tongue: Tried to do one commit at a time but failed.

File: user.js [hunk 1]
```diff
@@ -795,14 +797,13 @@ user_pref("security.pki.sha1_enforcement_level",		1);
 
 // PREF: Warn the user when server doesn't support RFC 5746 ("safe" renegotiation)
 // https://wiki.mozilla.org/Security:Renegotiation#security.ssl.treat_unsafe_negotiation_as_broken
-// see also CVE-2009-3555
+// https://web.nvd.nist.gov/view/vuln/detail?vulnId=CVE-2009-3555
 user_pref("security.ssl.treat_unsafe_negotiation_as_broken",	true);
 
-// PREF: ?? (disabled)
+// PREF: Disallow connection to servers not supporting safe renegotiation
```
Comment (pyllyukko) [inline user.js]: Shouldn't this have the "(disabled)" suffix, as the setting is still disabled.

File: README.md [hunk 2]
```diff
@@ -596,6 +596,12 @@ For more information, see [CONTRIBUTING](https://github.com/pyllyukko/user.js/bl
 
 --------------------------------------------------------------------------
 
+## License
```
Comment (pyllyukko) [inline README.md]: This is already quite evident at least when browsing the README through GitHub:

license

File: user.js [hunk 3]
```diff
@@ -506,19 +506,19 @@ user_pref("security.sri.enable",				true);
 // https://en.wikipedia.org/wiki/Do_not_track_header
 // https://dnt-dashboard.mozilla.org
 // https://github.com/pyllyukko/user.js/issues/11
-// NOTICE: DNT must be enabled manually
+// NOTICE: Do No Track must be enabled manually
```
Comment (pyllyukko) [inline user.js]: Could be "Do Not Track (DNT)"

File: user.js [hunk 4]
```diff
@@ -506,19 +506,19 @@ user_pref("security.sri.enable",				true);
...
-// NOTICE: Breaks visualisation of 3rd-party sites on the Lightbeam addon
+// NOTICE: Spoofing referers breaks functionality on websites relying on authentic referer headers
+// NOTICE: Spoofing referers breaks visualisation of 3rd-party sites on the Lightbeam addon
 // TODO: https://github.com/pyllyukko/user.js/issues/94, commented-out XOriginPolicy/XOriginTrimmingPolicy = 2 prefs
 user_pref("network.http.referer.spoofSource",			true);
 
 // PREF: Accept Only 1st Party Cookies
 // http://kb.mozillazine.org/Network.cookie.cookieBehavior#1
-// NOTICE: Breaks a number of payment gateways
+// NOTICE: Blocking 3rd-party cookies breaks a number of payment gateways
```
Comment (pyllyukko) [inline user.js]: We should have some refs to back this up

File: README.md [hunk 5]
```diff
@@ -372,8 +372,22 @@ See also:
 Hardening your often implies a trade-off with ease-of-use and comes with reduced functionality. Here is a list of known problems/limitations:
 
 <!-- BEGIN PROBLEMS-LIMITATIONS -->
-
-
+* Disabling DOM storage is known to cause`TypeError: localStorage is null` errors
+* IndexedDB could be used for tracking purposes, but is required for some add-ons to work (notably uBlock), so is left enabled
+* Firefox Hello requires setting `media.peerconnection.enabled` and `media.getusermedia.screensharing.enabled` to true, `security.OCSP.require` to false to work.
```
Comment (pyllyukko) [inline README.md]: Maybe we should get rid of this, as Firefox Hello is no more?

File: README.md [hunk 6]
```diff
@@ -372,8 +372,22 @@ See also:
...
+* Do No Track must be enabled manually
+* Spoofing referers breaks functionality on websites relying on authentic referer headers
+* Spoofing referers breaks visualisation of 3rd-party sites on the Lightbeam addon
+* Blocking 3rd-party cookies breaks a number of payment gateways
+* You can not view or inspect cookies when in private browsing: https://bugzilla.mozilla.org/show_bug.cgi?id=823941
+* Installing user.js will **remove your saved passwords** (https://github.com/pyllyukko/user.js/issues/27)
+* OCSP leaks your IP and domains you visit to the CA when OCSP Stapling is not available on visited host
+* OCSP is vulnerable to replay attacks when nonce is not configured on the OCSP responder
+* OCSP adds latency (performance)
+* Short-lived certificates are not checked for revocation (security.pki.cert_short_lifetime_in_days, default:10)
```
Comment (pyllyukko) [inline README.md]: Added #246

File: user.js [hunk 7]
```diff
@@ -511,11 +511,17 @@ user_pref("security.sri.enable",				true);
...
+// https://github.com/pyllyukko/user.js/issues/227
 // NOTICE: Spoofing referers breaks functionality on websites relying on authentic referer headers
 // NOTICE: Spoofing referers breaks visualisation of 3rd-party sites on the Lightbeam addon
+// NOTICE: Spoofing referers disables CSRF protection on some login pages not implementing origin-header/cookie+token based CSRF protection
 // TODO: https://github.com/pyllyukko/user.js/issues/94, commented-out XOriginPolicy/XOriginTrimmingPolicy = 2 prefs
 user_pref("network.http.referer.spoofSource",			true);
 
+// PREF: Don't send referer headers when following links across different domains
+// https://github.com/pyllyukko/user.js/issues/227
+user_pref("network.http.referer.XOriginPolicy",		2);
```
Comment (pyllyukko) [inline user.js]: See here: <https://github.com/pyllyukko/user.js/issues/227#issuecomment-287536761>

File: README.md [hunk 8]
```diff
@@ -369,13 +370,11 @@ See also:
 
 ## Known problems
 
-There are plenty! Hardening your browser will break your interwebs. Here's some examples:
+Hardening your often implies a trade-off with ease-of-use and comes with reduced functionality. Here is a list of known problems/limitations:
```
Comment (pyllyukko) [inline README.md]: "Hardening your"... what?

File: README.md [hunk 9]
```diff
@@ -372,8 +372,22 @@ See also:
...
-
+* Disabling DOM storage is known to cause`TypeError: localStorage is null` errors
+* IndexedDB could be used for tracking purposes, but is required for some add-ons to work (notably uBlock), so is left enabled
+* Firefox Hello requires setting `media.peerconnection.enabled` and `media.getusermedia.screensharing.enabled` to true, `security.OCSP.require` to false to work.
+* Do No Track must be enabled manually
+* Spoofing referers breaks functionality on websites relying on authentic referer headers
+* Spoofing referers breaks visualisation of 3rd-party sites on the Lightbeam addon
+* Blocking 3rd-party cookies breaks a number of payment gateways
+* You can not view or inspect cookies when in private browsing: https://bugzilla.mozilla.org/show_bug.cgi?id=823941
+* Installing user.js will **remove your saved passwords** (https://github.com/pyllyukko/user.js/issues/27)
```
Comment (pyllyukko) [inline README.md]: Maybe we should have similar refs parsing as in "what does it do" section?

**Comment by nodiscc** [comment]
Fixed all points mentioned in the review https://github.com/pyllyukko/user.js/pull/239/commits/f14e293fd485645876e2ef05180b3a768e1bddb5

**Pull request merged**